### ISSUE TITLE
Clang doesn't warn about always false comparisons

### ISSUE THREAD

**Issue by mushenoy** [llvm/llvm-project#83835]
Clang doesn't warn about always false comparisons. 
GCC reports the warning for the same program

Sample program:
```
#include <stddef.h>
#define getstr(s) ((const char *)((s) + 1))

int test(const char* s)
{
 if (s== NULL || getstr(s) == NULL)
 {
         return 1;
 }
 return 0;
}
```
Compilation with clang:
```
# clang -c -Wall program.c
#
```

Compilation with GCC:
```
# gcc -c -Wall program.c
program.c: In function test:
program.c:6:28: warning: the comparison will always evaluate as _false_ for the pointer operand in _s + 1_ must not be NULL [-Waddress]
    6 |  if (s== NULL || getstr(s) == NULL)
      |
```

Compiler versions:
```
clang version 16.0.6 
gcc (GCC) 13.2.1
```

**Comment by DimitryAndric** [comment]
clang has a different warning for this construct:

```text
% clang -O2 -Weverything -c program.c
program.c:4:5: warning: no previous prototype for function 'test' [-Wmissing-prototypes]
    4 | int test(const char* s)
      |     ^
program.c:4:1: note: declare 'static' if the function is not intended to be used outside of this translation unit
    4 | int test(const char* s)
      | ^
      | static
program.c:6:18: warning: unsafe pointer arithmetic [-Wunsafe-buffer-usage]
    6 |  if (s== NULL || getstr(s) == NULL)
      |                  ^~~~~~~~~
program.c:2:35: note: expanded from macro 'getstr'
    2 | #define getstr(s) ((const char *)((s) + 1))
      |                                   ^~~
2 warnings generated.
```

However, the `-Wunsafe-buffer-usage` warning does not seem to be enabled for either `-Wall` or `-Wextra`.

**Comment by pinskia** [comment]
I am not even sure how just doing `s+1` is unsafe either ...

**Comment by DimitryAndric** [comment]
> I am not even sure how just doing `s+1` is unsafe either ...

By itself it would be unsafe, if s could be NULL. But that was just disproven by the previous `s == NULL` check. So indeed, I have no idea where it gets this from. :)

**Comment by mushenoy** [comment]
clang seems to have a warning called: ```[-Wtautological-pointer-compare] ``` 

This warning is reported when compiling the following program, but not with above mentioned program.
```
#include <stdio.h>

struct ptr { int i; int j; };
int test (struct ptr *p)
{
    if (!p || p->i || (&p->j == NULL)) {
        return 1;
    }
    return 0;
}
```

Compilation with clang:
```
# clang -c -Wall  p2.c
p2.c:6:28: warning: comparison of address of 'p->j' equal to a null pointer is always false [-Wtautological-pointer-compare]
    if (!p || p->i || (&p->j == NULL)) {
                        ~~~^    ~~~~
1 warning generated.
```
Shouldn't this warning be reported for the program mentioned in the issue as well?

**Comment by cor3ntin** [comment]
> Shouldn't this warning be reported for the program mentioned in the issue as well?

Ideally yes
